Core/include/Acts/TrackFitting/detail/GsfActor.hpp

```diff
@@ -0,0 +1,728 @@
+
+#pragma once
+
+#include "Acts/EventData/MultiComponentBoundTrackParameters.hpp"
+#include "Acts/EventData/MultiTrajectory.hpp"
+#include "Acts/EventData/MultiTrajectoryHelpers.hpp"
+#include "Acts/MagneticField/MagneticFieldProvider.hpp"
+#include "Acts/Material/ISurfaceMaterial.hpp"
+#include "Acts/Surfaces/CylinderSurface.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/TrackFitting/GsfError.hpp"
+#include "Acts/TrackFitting/KalmanFitter.hpp"
+#include "Acts/TrackFitting/detail/BetheHeitlerApprox.hpp"
+#include "Acts/TrackFitting/detail/GsfSmoothing.hpp"
+#include "Acts/TrackFitting/detail/GsfUtils.hpp"
+#include "Acts/TrackFitting/detail/KLMixtureReduction.hpp"
+#include "Acts/TrackFitting/detail/KalmanUpdateHelpers.hpp"
+#include "Acts/Utilities/Zip.hpp"
+
+#include <ios>
+#include <map>
+#include <numeric>
+
+namespace Acts {
+namespace detail {
+
+struct GsfResult {
+  /// The multi-trajectory which stores the graph of components
+  MultiTrajectory fittedStates;
+  
+  /// This provides the weights for the states in the MultiTrajectory. Each entry maps to one track state. TODO This is a workaround until the MultiTrajectory can handle weights
+  std::map<size_t, ActsScalar> weightsOfStates;
+
+  /// The current indexes for the newest components in the multi trajectory
+  /// (this includes material, hole and outlier states)
+  std::vector<std::size_t> currentTips;
+
+  /// The last tips referring to a measuerement state so we do not need so
+  /// search them recursively later
+  std::vector<std::size_t> lastMeasurementTips;
+
+  /// We must capture the parent tips to ensure that we can keep track of the
+  /// last states in the multitrajectory after the component convolution and
+  /// reduction
+  std::vector<std::size_t> parentTips;
+
+  /// Some counting
+  std::size_t measurementStates = 0;
+  std::size_t measurementHoles = 0;
+  std::size_t processedStates = 0;
+  std::set<Acts::GeometryIdentifier> visitedSurfaces;
+  std::vector<const Acts::Surface*> missedActiveSurfaces;
+
+  // Propagate potential errors to the outside
+  Result<void> result{Result<void>::success()};
+
+  // Used for workaround to initialize MT correctly
+  bool haveInitializedResult = false;
+};
+
+/// The actor carrying out the GSF algorithm
+template <typename bethe_heitler_approx_t>
+struct GsfActor {
+  /// Enforce default construction
+  GsfActor() = default;
+
+  /// Broadcast the result_type
+  using result_type = GsfResult;
+
+  // Actor configuration
+  struct Config {
+    /// Maximum number of components which the GSF should handle
+    std::size_t maxComponents = 16;
+
+    /// Input measurements
+    std::map<GeometryIdentifier, std::reference_wrapper<const SourceLink>>
+        inputMeasurements;
+
+    /// Bethe Heitler Approximator pointer. The fitter holds the approximator
+    /// instance TODO if we somehow could initialize a reference here...
+    const bethe_heitler_approx_t* bethe_heitler_approx = nullptr;
```

Comment:
Could that be a case we use Paul's `Delegate<>()` class?

Reply:
Hmm not sure if  this is that easy, since the return type of the `.mixture()` function depends on the `NComponents` integer-template-parameter (it is a `std::array<..., NComponents>`), and the `Delegate` needs a stable return-type, right?
I also don't think this template instatiation is expensive...